src/storage/lsm.rs

```diff
@@ -69,98 +57,158 @@ impl State {
+            SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_micros()));
+        let sstable = write_sstable(&self.memtable, path)?;
+        self.sstables.push(sstable);
+        self.memtable.0.clear();
+        self.commit_log.take(); // Close the file.
+        self.commit_log = Some(
+            OpenOptions::new()
+                .create(true)
+                .write(true)
+                .append(true)
+                .open(COMMIT_LOG_PATH)?,
+        );
+
+        Ok(())
+    }
+}
 
+/// Read all key-value pairs.
+fn read_commit_log(path: &PathBuf) -> Memtable {
     let mut memtable = Memtable::default();
-    loop {
-        if let Ok((key, val)) = get_key_value() {
-            memtable.0.insert(key, val);
-        } else {
-            break;
+
+    let file_result = File::open(path);
+    if let Ok(mut file) = file_result {
+        let iter = serde::KeyValueIterator { file: &mut file };
+        for (_, key, maybe_val) in iter {
+            match maybe_val {
+                None => {
+                    memtable.0.remove(&key);
+                }
+                Some(val) => {
+                    memtable.0.insert(key, val);
+                }
+            }
         }
     }
-    Ok(memtable)
+    // Else, ignore if commit log does not exist.
+
+    memtable
 }
 
-fn append_to_commit_log(file: &mut File, k: &Key, v: &Option<Value>) -> Result<()> {
-    file.write(&k.0.len().to_le_bytes())?;
-    file.write(k.0.as_bytes())?;
-    match v {
-        Some(Value::Bytes(v)) => {
-            file.write(&v.len().to_le_bytes())?;
-            file.write(v)?;
-        }
-        _ => {
-            let zero: usize = 0;
-            file.write(&zero.to_le_bytes())?;
-        }
+fn write_sstable(memtable: &Memtable, path: PathBuf) -> Result<SSTable> {
+    let mut file = OpenOptions::new()
+        .create(true)
+        .write(true)
+        .truncate(true)
+        .open(&path)?;
+
+    for kv in memtable.0.iter() {
+        serde::write_kv(kv.0, Some(&kv.1), &mut file)?;
     }
-    Ok(())
+
+    Ok(read_sstable(path)?)
 }
 
-fn read_sstable(file: File) -> SSTable {
-    // TODO
-    let idx = BTreeMap::<Key, SeekFrom>::new();
-    SSTable {
-        file,
-        idx
+fn read_sstable(path: PathBuf) -> Result<SSTable> {
+    let mut idx = BTreeMap::<Key, u64>::new();
+
+    let mut offset = 0usize;
+
+    let mut file = File::open(&path)?;
+    let iter = serde::KeyValueIterator { file: &mut file };
+    for (delta_offset, key, _) in iter.step_by(SSTABLE_IDX_SPARSENESS) {
+        idx.insert(key, offset as u64);
+        offset += delta_offset;
     }
+
+    Ok(SSTable { path, idx })
 }
 
-fn search_in_sstable(ss: &SSTable, k: &Key) -> Option<Value> {
-    // TODO
-    // 1. bisect in ss.idx
-    // 1. Seek linearlly in file
-    None
+/// 1. Bisect in the in-memory sparse index.
+/// 1. Seek linearlly in file.
+fn search_in_sstable(ss: &SSTable, k: &Key) -> Result<Option<Value>> {
```

Comment:
do you have a preference toward free functions versus `impl` methods? 

I'm seeing that this could also be written:

``` rust
impl SSTable {
  fn search(&self, k: &Key) -> Result<...> {
  }
}
```

Reply:
Good idea. Moved other fns too.